src/analysis.ts

```diff
@@ -12,16 +12,309 @@
+    for (const descriptor of path) {
+      if (descriptor instanceof DocumentDescriptor) {
+        pathToElement = descriptor.document.url;
+      }
+    }
+    if (!pathToElement) {
+      throw new Error(`Unable to determine path to element: ${element}`);
+    }
+    if (this.elementPaths.has(element)) {
+      if (this.elementPaths.get(element) !== pathToElement) {
+        throw new Error(
+            `Found element ${element} at distinct paths: ` +
+            `${pathToElement} and ${this.elementPaths.get(element)}`);
+      }
+    } else {
+      this.elementPaths.set(element, pathToElement);
+      const elem = serializeElementDescriptor(element, pathToElement);
+      if (elem) {
+        this.elements.push(elem);
+      }
+    }
+  }
+}
+
+abstract class AnalysisVisitor {
+  visitDocumentDescriptor?(dd: DocumentDescriptor, path: Descriptor[]): void;
+  visitInlineDocumentDescriptor?
+      (dd: InlineDocumentDescriptor<any>, path: Descriptor[]): void;
+  visitElement?(element: ElementDescriptor, path: Descriptor[]): void;
+  visitDocument?(document: Document<any, any>, path: Descriptor[]): void;
+  done?(): void;
+}
+
+class AnalysisWalker {
+  private descriptors: DocumentDescriptor[];
+  private path: Descriptor[] = [];
+  constructor(descriptors: DocumentDescriptor[]) {
+    this.descriptors = descriptors;
+  }
+  walk(visitors: AnalysisVisitor[]) {
+    this.path.length = 0;
+    for (const descriptor of this.descriptors) {
+      this._walkDocumentDescriptor(descriptor, visitors);
+    }
+    for (const visitor of visitors) {
+      if (visitor.done) {
+        visitor.done();
+      }
+    }
+  }
+
+  private _walkDocumentDescriptor(
+      dd: DocumentDescriptor, visitors: AnalysisVisitor[]) {
+    this.path.push(dd);
+
+    for (const visitor of visitors) {
+      if (visitor.visitDocumentDescriptor) {
+        visitor.visitDocumentDescriptor(dd, this.path);
+      }
+    }
+
+    for (const entity of dd.entities) {
+      this._walkEntity(entity, visitors);
+    }
+    for (const dependency of dd.dependencies) {
+      this._walkEntity(dependency, visitors);
+    }
+    this._walkDocument(dd.document, visitors);
+    this.path.pop();
+  }
+
+  private _walkInlineDocumentDescriptor(
+      dd: InlineDocumentDescriptor<any>, visitors: AnalysisVisitor[]) {
+    for (const visitor of visitors) {
+      if (visitor.visitInlineDocumentDescriptor) {
+        visitor.visitInlineDocumentDescriptor(dd, this.path);
+      }
+    }
+  }
+
+  private _walkEntity(entity: Descriptor, visitors: AnalysisVisitor[]) {
+    if (entity instanceof DocumentDescriptor) {
+      return this._walkDocumentDescriptor(entity, visitors);
+    } else if (entity instanceof InlineDocumentDescriptor) {
+      return this._walkInlineDocumentDescriptor(entity, visitors);
+    } else if (entity['type'] === 'element') {
+      return this._walkElement(<ElementDescriptor>entity, visitors);
+    }
+    throw new Error(`Unknown kind of descriptor: ${entity}`);
```

Comment:
Are there any other kinds of descriptors that might contain other descriptors? I wanted the walk to have at least a chance of being complete before I allowed an open world.

Is there a better way to write a descriptor visitor?